### ISSUE TITLE
[Developer-Experience] Adding pre-commit hooks or tox for consistency while creating PRs

### ISSUE THREAD

**Issue by vivekkhimani** [skypilot-org/skypilot#1186]
## Tl;dr
I am a new contributor to this repo and have been working on one of the first issues. I was trying to run the ```yapf``` formatting, tests, and ```PyLint``` before pushing a PR for review. I noticed that workflows exist for each of these tools which get triggered while merging to master or creating a release.

## Proposal
I am proposing adding ```tox``` pipeline and ```pre-commit```hooks (which are automatically executed locally before a developer pushes the code to their branches). This will ensure that all the essential checks are performed and the code is cleaned before it's pushed for review (and merged).

While ```pre-commit``` can be used to run ```yapf``` based formatting and ```pylint```, ```tox``` can be used to run ```pytest``` and other related utilities (in the future).

- Pre-commit docs: https://pre-commit.com/
- Example config: https://github.com/vivekkhimani/torchfl/blob/master/.pre-commit-config.yaml
- tox docs: https://tox.wiki/en/latest/

## Changes/Impact
Once the pre-commit and tox configs are added, the hooks will be created in the local developer environment. When the developer tries to commit the code, these hooks will be executed, and run formatters, linters on the repo before the code can be pushed. Devs can also run pre-commit manually by calling: ```pre-commit run --all```. ```tox``` can be used by the developers to run the test cases before pushing the code.

**Comment by romilbhardwaj** [comment]
Thanks for this suggestion @vivekkhimani! I think this is indeed a step in the right direction and we should have some kind of automated client-side checks. Some thoughts:

1. The linter and the tests (other than the smoke tests) can take some time to run. We should probably run them only pre-push, and have an easy way to bypass the checks if the author wants (`--no-verify` flag from git should work).
2. Can we use regular git hooks instead of `pre-commit` and `tox`? There are indeed benefits to using `pre-commit` and `tox`, but at the current stage of this project, I feel it would be good to (a) keep external dependencies to a minimum and (b) retain as much flexibility as we can.

What do you think? I'll also let others chime in. 

If it's not too hard to add and has a net positive impact on dev experience, I feel we should pursue this direction.

**Comment by vivekkhimani** [comment]
Thanks, for a quick response @romilbhardwaj! The direction I was having essential checks that can be run super quickly in ```pre-commit``` and have ```tox``` steps defined for tests and linters. As developers have more control over ```tox``` they can choose to run it if they have made any changes to those files.

Also, the reason I proposed ```pre-commit``` over git hooks as it's sort of another layer of abstraction over git hooks. Yes, it will be an additional dependency but a lot of their pre-built hooks provide a lot of value.

I can spin a sample workflow and we can all test it, discuss, and see if it's worth adding? How does that sound? I will spin up a quick PR soon.

**Comment by romilbhardwaj** [comment]
Sounds good! A PR would be very helpful in evaluating this decision. Thanks @vivekkhimani!

**Comment by Michaelvll** [comment]
@vivekkhimani, just a kind check if you still want to have a PR for this issue? : )
